commit c0ac6c05e04e43e16c8a626a27d09b5b925cdb31
Author: Jinwoo Jeong <jjw8967@ajou.ac.kr>
Date:   Tue Jul 23 04:53:26 2024 +0000

    Implement the functions of DeepPlan

diff --git a/aten/src/ATen/cuda/CachingHostAllocator.cpp b/aten/src/ATen/cuda/CachingHostAllocator.cpp
index a6646779cfb..ad48f1cfc38 100644
--- a/aten/src/ATen/cuda/CachingHostAllocator.cpp
+++ b/aten/src/ATen/cuda/CachingHostAllocator.cpp
@@ -383,7 +383,7 @@ void CachingHostAllocator_emptyCache() {
   getCUDAHostAllocator().empty_cache();
 }
 
-struct CUDAHostAllocatorWrapper final : public at::Allocator {
+struct HostAllocatorWrapper final : public at::Allocator {
   at::DataPtr allocate(size_t size) const override {
     auto ptr_and_ctx = getCUDAHostAllocator().allocate(size);
     return {
@@ -394,9 +394,29 @@ struct CUDAHostAllocatorWrapper final : public at::Allocator {
   }
 };
 
-static CUDAHostAllocatorWrapper cuda_host_allocator;
+static HostAllocatorWrapper host_allocator;
 
 at::Allocator* getCachingHostAllocator() {
+  return &host_allocator;
+}
+
+struct CUDAHostAllocatorWrapper final : public at::Allocator {
+  at::DataPtr allocate(size_t size) const override {
+    int device;
+    C10_CUDA_CHECK(cudaGetDevice(&device));
+    auto ptr_and_ctx = getCUDAHostAllocator().allocate(size);
+
+    return {
+        ptr_and_ctx.first,
+        ptr_and_ctx.second,
+        &CUDAHostAllocatorDeleter,
+        at::Device(at::DeviceType::CUDA, device)};
+  }
+};
+
+static CUDAHostAllocatorWrapper cuda_host_allocator;
+
+at::Allocator* getCachingCUDAHostAllocator() {
   return &cuda_host_allocator;
 }
 
diff --git a/aten/src/ATen/cuda/CachingHostAllocator.h b/aten/src/ATen/cuda/CachingHostAllocator.h
index 65ad7f7d16e..bcf6194f1ba 100644
--- a/aten/src/ATen/cuda/CachingHostAllocator.h
+++ b/aten/src/ATen/cuda/CachingHostAllocator.h
@@ -21,6 +21,7 @@ namespace at::cuda {
 // blocks, unlike the caching device allocator.
 //
 TORCH_CUDA_CPP_API c10::Allocator* getCachingHostAllocator();
+TORCH_CUDA_CPP_API c10::Allocator* getCachingCUDAHostAllocator();
 
 // Records an event in the specified stream. The allocation corresponding to the
 // input `ptr`/`ctx` will not be re-used until the event has occurred.
diff --git a/aten/src/ATen/cuda/PinnedMemoryAllocator.cpp b/aten/src/ATen/cuda/PinnedMemoryAllocator.cpp
index 973027cd87f..e2242a03f26 100644
--- a/aten/src/ATen/cuda/PinnedMemoryAllocator.cpp
+++ b/aten/src/ATen/cuda/PinnedMemoryAllocator.cpp
@@ -29,4 +29,24 @@ Tensor _pin_memory_cuda(const Tensor& self, c10::optional<Device> device) {
 }
 
 
+Tensor _cuda_host_cuda(const Tensor& self, c10::optional<Device> device) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!device.has_value());
+  if (self.is_pinned() && self.device().is_cuda()) {
+    TORCH_WARN("Already '", self.toString(), "' is pinned");
+    return self;
+  }
+  auto* allocator = at::cuda::getCUDAHostMemoryAllocator();
+  auto storage = Storage(
+      Storage::use_byte_size_t(),
+      detail::computeStorageNbytes(
+        self.sizes(), self.strides(), self.dtype().itemsize()),
+      allocator,
+      /*resizable=*/false);
+
+  auto tensor = at::empty({0}, self.options().device(at::kCUDA)).set_(storage, 0, self.sizes(), self.strides());
+  tensor.copy_(self);
+  return tensor;
+}
+
+
 } // namespace at::native
diff --git a/aten/src/ATen/cuda/PinnedMemoryAllocator.h b/aten/src/ATen/cuda/PinnedMemoryAllocator.h
index 854f5d8dd12..ff337867647 100644
--- a/aten/src/ATen/cuda/PinnedMemoryAllocator.h
+++ b/aten/src/ATen/cuda/PinnedMemoryAllocator.h
@@ -8,4 +8,8 @@ namespace at::cuda {
 inline TORCH_CUDA_CPP_API at::Allocator* getPinnedMemoryAllocator() {
   return getCachingHostAllocator();
 }
+
+inline TORCH_CUDA_CPP_API at::Allocator* getCUDAHostMemoryAllocator() {
+  return getCachingCUDAHostAllocator();
+}
 } // namespace at::cuda
diff --git a/aten/src/ATen/cuda/TensorSync.cpp b/aten/src/ATen/cuda/TensorSync.cpp
new file mode 100644
index 00000000000..93b985b0f31
--- /dev/null
+++ b/aten/src/ATen/cuda/TensorSync.cpp
@@ -0,0 +1,138 @@
+#include <ATen/core/Tensor.h>
+#include <ATen/cuda/TensorSync.h>
+#include <c10/cuda/CUDAStream.h>
+
+#include <cuda_runtime_api.h>
+#include <map>
+#include <mutex>
+#include <condition_variable>
+#include <iostream>
+
+namespace at {
+namespace cuda::EventManager {
+
+struct DeviceEventManager {
+ public:
+  DeviceEventManager(int device) : device(device) {};
+
+  void recordEvent(at::Tensor& src, const at::Tensor& dst, at::cuda::CUDAStream stream) {
+    {
+      std::lock_guard<std::mutex> lock(mutex);
+      void *src_ptr = src.data_ptr();
+      void *dst_ptr;
+
+      src.set_data(dst);
+
+      dst_ptr = src.data_ptr();
+
+      cudaEvent_t event = createCudaEvent();
+      C10_CUDA_CHECK(cudaEventRecord(event, stream));
+
+      // Deletes items that already exist.
+      auto it = recorded_event_map.find(src_ptr);
+      recorded_event_map[src_ptr] = {dst_ptr, event};
+
+      it = recorded_event_map.find(dst_ptr);
+      recorded_event_map[dst_ptr] = {src_ptr, event};
+
+    }
+    cv.notify_all();
+  }
+
+  void syncEvent(void* ptr, at::cuda::CUDAStream stream, bool same_device) {
+    cudaEvent_t event;
+    {
+      std::unique_lock<std::mutex> lock(mutex);
+      auto it = recorded_event_map.find(ptr);
+      if (it == recorded_event_map.end()) {
+        if (same_device) {
+          return;
+        }
+        cv.wait(lock, [&] {
+            auto _it = this->recorded_event_map.find(ptr);
+            return _it != this->recorded_event_map.end();
+            });
+        it = recorded_event_map.find(ptr);
+      }
+
+      event = it->second.second;
+      void* linked_ptr = it->second.first;
+      recorded_event_map.erase(it);
+      it = recorded_event_map.find(linked_ptr);
+      if (it != recorded_event_map.end()) {
+        recorded_event_map.erase(it);
+      }
+    }
+
+    //C10_CUDA_CHECK(cudaEventSynchronize(event));
+    //freeCudaEvent(event);
+    C10_CUDA_CHECK(cudaStreamWaitEvent(stream, event, 0));
+  }
+
+  cudaEvent_t createCudaEvent() {
+    cudaEvent_t event;
+    C10_CUDA_CHECK(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+
+    return event;
+  }
+
+  void freeCudaEvent(cudaEvent_t event) {
+    C10_CUDA_CHECK(cudaEventDestroy(event));
+  }
+
+  std::mutex mutex;
+  std::condition_variable cv;
+  int device;
+  std::map<void*, std::pair<void*, cudaEvent_t>> recorded_event_map;
+};
+
+struct EventManager {
+ public:
+  void init(int device_count) {
+    int size = device_event_managers.size();
+    if (size < device_count) {
+      device_event_managers.resize(device_count);
+      for (int i = 0; i < device_count; i++) {
+        device_event_managers[i] = std::unique_ptr<DeviceEventManager>(
+            new DeviceEventManager(i));
+      }
+    }
+  }
+
+  void recordEvent(at::Tensor& src, const at::Tensor& dst, int device, at::cuda::CUDAStream stream) {
+    device_event_managers[device]->recordEvent(src, dst, stream);
+  }
+
+  void syncEvent(void* ptr, int device, int cur_device, at::cuda::CUDAStream stream) {
+    device_event_managers[device]->syncEvent(ptr, stream, device == cur_device);
+  }
+
+  std::vector<std::unique_ptr<DeviceEventManager>> device_event_managers;
+};
+
+static EventManager event_manager;
+
+void initEventManager(int device_count) {
+  event_manager.init(device_count);
+}
+
+}  // namespace at::cuda::EventManager
+
+namespace native {
+
+void _record_event_cuda(at::Tensor& self, const at::Tensor& tensor, c10::Device device) {
+  if (self.data_ptr() == tensor.data_ptr()) {
+    return;
+  }
+
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  cuda::EventManager::event_manager.recordEvent(self, tensor, device.index(), stream);
+}
+
+void _sync_event_cuda(at::Tensor& self, c10::Device device, c10::Device cur_device) {
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  cuda::EventManager::event_manager.syncEvent(self.data_ptr(), device.index(), cur_device.index(), stream);
+}
+
+}  // namespace at::native
+}  // namespace at
diff --git a/aten/src/ATen/cuda/TensorSync.h b/aten/src/ATen/cuda/TensorSync.h
new file mode 100644
index 00000000000..f85b769d5a1
--- /dev/null
+++ b/aten/src/ATen/cuda/TensorSync.h
@@ -0,0 +1,11 @@
+#pragma oncea
+
+#include <c10/cuda/CUDAStream.h>
+#include <ATen/core/Tensor.h>
+
+namespace at::cuda::EventManager {
+
+TORCH_CUDA_CPP_API void
+initEventManager(int device_count);
+
+}  // namespace at::cuda
diff --git a/aten/src/ATen/cuda/detail/CUDAHooks.cpp b/aten/src/ATen/cuda/detail/CUDAHooks.cpp
index acb9b1931f0..ae279444390 100644
--- a/aten/src/ATen/cuda/detail/CUDAHooks.cpp
+++ b/aten/src/ATen/cuda/detail/CUDAHooks.cpp
@@ -11,6 +11,7 @@
 #include <ATen/cuda/PeerToPeerAccess.h>
 #include <ATen/cuda/PinnedMemoryAllocator.h>
 #include <ATen/cuda/nvrtc_stub/ATenNVRTC.h>
+#include <ATen/cuda/TensorSync.h>
 #include <ATen/detail/CUDAHooksInterface.h>
 #include <ATen/native/cuda/CuFFTPlanCache.h>
 #include <c10/util/Exception.h>
@@ -107,6 +108,8 @@ void CUDAHooks::initCUDA() const {
   c10::cuda::CUDACachingAllocator::init(num_devices);
   at::cuda::detail::init_p2p_access_cache(num_devices);
 
+  at::cuda::EventManager::initEventManager(num_devices);
+
 #if AT_MAGMA_ENABLED()
   TORCH_INTERNAL_ASSERT(magma_init_fn != nullptr, "Cannot initialize magma, init routine not set");
   magma_init_fn();
diff --git a/aten/src/ATen/native/Memory.cpp b/aten/src/ATen/native/Memory.cpp
index 2b66f089339..467dc4991cb 100644
--- a/aten/src/ATen/native/Memory.cpp
+++ b/aten/src/ATen/native/Memory.cpp
@@ -8,8 +8,10 @@
 #else
 #include <ATen/ops/_debug_has_internal_overlap_native.h>
 #include <ATen/ops/_pin_memory.h>
+#include <ATen/ops/_cuda_host.h>
 #include <ATen/ops/is_pinned_native.h>
 #include <ATen/ops/pin_memory_native.h>
+#include <ATen/ops/cuda_host_native.h>
 #endif
 
 namespace at {
@@ -37,5 +39,12 @@ Tensor pin_memory(const Tensor& self, c10::optional<Device> device) {
   return at::_pin_memory(self, device);
 }
 
+Tensor cuda_host(const Tensor& self, c10::optional<Device> device) {
+  if (self.is_pinned() && self.device().is_cuda()) {
+    return self;
+  }
+  return at::_cuda_host(self, device);
+}
+
 }
 }
diff --git a/aten/src/ATen/native/TensorSync.cpp b/aten/src/ATen/native/TensorSync.cpp
new file mode 100644
index 00000000000..e1cbe6c8923
--- /dev/null
+++ b/aten/src/ATen/native/TensorSync.cpp
@@ -0,0 +1,41 @@
+#include <ATen/core/Tensor.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_record_event.h>
+#include <ATen/ops/_sync_event.h>
+#include <ATen/ops/_record_event_native.h>
+#include <ATen/ops/_sync_event_native.h>
+#include <ATen/ops/is_pinned_native.h>
+#endif
+
+namespace at::native {
+
+void record_tensor(Tensor& self, const Tensor& tensor, Device device) {
+  if (self.is_cuda() && self.is_pinned())
+    return;
+
+  at::_record_event(self, tensor, device);
+}
+
+Tensor& sync_device_(Tensor& self, Device device) {
+  if (self.is_cuda() && self.is_pinned())
+    return self;
+
+  at::_sync_event(self, device, self.device());
+  return self;
+}
+
+Tensor& sync_tensor_(Tensor& self, const Tensor& tensor) {
+  if (self.is_cuda() && self.is_pinned())
+    return self;
+  if (!tensor.is_cuda())
+    return self;
+
+  at::_sync_event(self, tensor.device(), self.device());
+  return self;
+}
+
+}  // namespace at::native
diff --git a/aten/src/ATen/native/cuda/Copy.cu b/aten/src/ATen/native/cuda/Copy.cu
index 3525d85f513..d4b616bfd27 100644
--- a/aten/src/ATen/native/cuda/Copy.cu
+++ b/aten/src/ATen/native/cuda/Copy.cu
@@ -77,12 +77,14 @@ void copy_device_to_device(TensorIterator& iter,
     // write-after-read dependencies on the destination side are handled, so
     // that no one is operating on the dst memory when we perform the copy.
     // src waits on dst barrier (src already waits on src)
-    CUDAEvent dst_ready;
-    device_guard.set_device(dst_device);
-    dst_ready.record(getCurrentCUDAStream(dst_device.index()));
+    if (!non_blocking) {
+      CUDAEvent dst_ready;
+      device_guard.set_device(dst_device);
+      dst_ready.record(getCurrentCUDAStream(dst_device.index()));
 
-    device_guard.set_device(src_device);
-    dst_ready.block(copy_stream);
+      device_guard.set_device(src_device);
+      dst_ready.block(copy_stream);
+    }
   }
 
   if (memcpy_eligible) {
@@ -121,11 +123,13 @@ void copy_device_to_device(TensorIterator& iter,
     // operate on dst's copy until the copy is complete.
 
     // Still on src_device, record stream event
-    CUDAEvent src_ready;
-    src_ready.record(copy_stream);
+    if (!non_blocking) {
+      CUDAEvent src_ready;
+      src_ready.record(copy_stream);
 
-    device_guard.set_device(dst_device);
-    src_ready.block(getCurrentCUDAStream(dst_device.index()));
+      device_guard.set_device(dst_device);
+      src_ready.block(getCurrentCUDAStream(dst_device.index()));
+    }
   }
 
   AT_CUDA_CHECK(cudaGetLastError());
diff --git a/aten/src/ATen/native/native_functions.yaml b/aten/src/ATen/native/native_functions.yaml
index b1658fb125b..2285a04ec7b 100644
--- a/aten/src/ATen/native/native_functions.yaml
+++ b/aten/src/ATen/native/native_functions.yaml
@@ -4354,6 +4354,34 @@
     MPS: _pin_memory_mps
   autogen: _pin_memory.out
 
+- func: cuda_host(Tensor(a) self, Device? device=None) -> Tensor(a)
+  variants: method
+
+- func: _cuda_host(Tensor self, Device? device=None) -> Tensor
+  dispatch:
+    CUDA: _cuda_host_cuda
+
+- func: record_tensor(Tensor(a!) self, Tensor old, Device device) -> ()
+  variants: function, method
+
+- func: sync_device_(Tensor(a!) self, Device device) -> Tensor(a!)
+  variants: function, method
+
+- func: sync_tensor_(Tensor(a!) self, Tensor tensor) -> Tensor(a!)
+  variants: function, method
+
+- func: _record_event(Tensor(a!) self, Tensor tensor, Device device) -> ()
+  variants: function, method
+  device_check: NoCheck
+  dispatch:
+    CUDA: _record_event_cuda
+
+- func: _sync_event(Tensor(a!) self, Device device, Device cur_deivce) -> ()
+  variants: function, method
+  device_check: NoCheck
+  dispatch:
+    CUDA: _sync_event_cuda
+
 - func: pinverse(Tensor self, float rcond=1e-15) -> Tensor
   variants: function, method
 
diff --git a/aten/src/ATen/templates/RegisterBackendSelect.cpp b/aten/src/ATen/templates/RegisterBackendSelect.cpp
index 6463701a493..9bd181c1641 100644
--- a/aten/src/ATen/templates/RegisterBackendSelect.cpp
+++ b/aten/src/ATen/templates/RegisterBackendSelect.cpp
@@ -12,6 +12,9 @@
 #include <ATen/Operators.h>
 #else
 #include <ATen/ops/is_pinned_ops.h>
+#include <ATen/ops/_cuda_host_ops.h>
+#include <ATen/ops/_record_event_ops.h>
+#include <ATen/ops/_sync_event_ops.h>
 #include <ATen/ops/_pin_memory_ops.h>
 
 ${ops_headers}
@@ -39,10 +42,28 @@ at::Tensor _pin_memory(const Tensor& self, c10::optional<at::Device> device) {
   return at::_ops::_pin_memory::redispatch(_dk, self, device);
 }
 
+at::Tensor _cuda_host(const Tensor& self, c10::optional<at::Device> device) {
+  DispatchKeySet _dk = c10::DispatchKeySet(c10::computeDispatchKey(c10::nullopt, self.layout(), at::kCUDA));
+  return at::_ops::_cuda_host::redispatch(_dk, self, device);
+}
+
+void _record_event(Tensor& self, const Tensor& tensor, at::Device device) {
+  DispatchKeySet _dk = c10::DispatchKeySet(c10::computeDispatchKey(c10::nullopt, self.layout(), at::kCUDA));
+  at::_ops::_record_event::redispatch(_dk, self, tensor, device);
+}
+
+void _sync_event(Tensor& self, at::Device src, at::Device dst) {
+  DispatchKeySet _dk = c10::DispatchKeySet(c10::computeDispatchKey(c10::nullopt, self.layout(), at::kCUDA));
+  at::_ops::_sync_event::redispatch(_dk, self, src, dst);
+}
+
 TORCH_LIBRARY_IMPL(aten, BackendSelect, m) {
   ${backend_select_function_registrations};
   m.impl(TORCH_SELECTIVE_NAME("aten::is_pinned"), TORCH_FN(is_pinned));
   m.impl(TORCH_SELECTIVE_NAME("aten::_pin_memory"), TORCH_FN(_pin_memory));
+  m.impl(TORCH_SELECTIVE_NAME("aten::_cuda_host"), TORCH_FN(_cuda_host));
+  m.impl(TORCH_SELECTIVE_NAME("aten::_record_event"), TORCH_FN(_record_event));
+  m.impl(TORCH_SELECTIVE_NAME("aten::_sync_event"), TORCH_FN(_sync_event));
 }
 
 } // namespace
diff --git a/torch/csrc/jit/api/module.cpp b/torch/csrc/jit/api/module.cpp
index 19f67e2f54d..40895de8f38 100644
--- a/torch/csrc/jit/api/module.cpp
+++ b/torch/csrc/jit/api/module.cpp
@@ -165,6 +165,91 @@ void Module::to(at::Device device, bool non_blocking) {
   to_impl(device, /*dtype=*/c10::nullopt, non_blocking);
 }
 
+void Module::cuda_host() {
+  for (at::Tensor e : parameters()) {
+    auto new_data = e.cuda_host();
+    e.set_data(new_data);
+  }
+  for (at::Tensor e : buffers()) {
+    auto new_data = e.cuda_host();
+    e.set_data(new_data);
+  }
+}
+
+void Module::to_and_record(at::Device device, bool non_blocking) {
+  //TODO: device must be cuda, otherwise return
+  c10::optional<at::Tensor> record_tensor;
+  for (auto e : named_parameters()) {
+    if (e.name.find("weight") != std::string::npos) {
+      record_tensor = e.value;
+    }
+    else {
+      auto new_data = e.value.to(device, non_blocking);
+      e.value.set_data(new_data);
+    }
+  }
+  for (auto e : named_buffers()) {
+    auto new_data = e.value.to(device, non_blocking);
+    e.value.set_data(new_data);
+  }
+
+  if (record_tensor.has_value()) {
+    auto new_data = record_tensor.value().to(device, non_blocking);
+    record_tensor.value().record_tensor(new_data, device);
+  }
+}
+
+void Module::synchronize(at::Device device) {
+  //TODO: device must be cuda, otherwise return
+  at::Tensor sync_tensor;
+  for (auto e : named_parameters()) {
+    if (e.name.find("weight") != std::string::npos) {
+      e.value.sync_device_(device);
+    }
+  }
+}
+
+void Module::pin_memory() {
+  for (at::Tensor e : parameters()) {
+    auto new_data = e.pin_memory();
+    e.set_data(new_data);
+  }
+  for (at::Tensor e : buffers()) {
+    auto new_data = e.pin_memory();
+    e.set_data(new_data);
+  }
+}
+
+// TODO Remove redundancy
+void Module::cuda_backup(bool non_blocking) {
+  for (auto named_param : named_parameters()) {
+    std::string name = "prev_" + named_param.name;
+    at::Tensor t = named_param.value.tensor_data();
+    register_attribute(name, TensorType::get(), t);
+  }
+  for (auto named_buf : named_buffers()) {
+    std::string name = "prev_" + named_buf.name;
+    at::Tensor t = named_buf.value.tensor_data();
+    register_attribute(name, TensorType::get(), t);
+  }
+}
+
+// TODO Remove redundancy
+void Module::clear() {
+  for (auto named_param : named_parameters()) {
+    std::string name = "prev_" + named_param.name;
+    auto t = _ivalue()->getAttr(name).toTensor();
+
+    named_param.value.set_data(t);
+  }
+  for (auto named_buf : named_buffers()) {
+    std::string name = "prev_" + named_buf.name;
+    auto t = _ivalue()->getAttr(name).toTensor();
+
+    named_buf.value.set_data(t);
+  }
+}
+
 static void module_state_to(
     const autograd::Variable& variable,
     const c10::optional<at::Device>& device,
diff --git a/torch/csrc/jit/api/module.h b/torch/csrc/jit/api/module.h
index 570ce12d1fa..4053c8a92d5 100644
--- a/torch/csrc/jit/api/module.h
+++ b/torch/csrc/jit/api/module.h
@@ -214,6 +214,18 @@ struct TORCH_API Module : public Object {
   /// effect.
   void to(at::Device device, bool non_blocking = false);
 
+  void cuda_host();
+
+  void to_and_record(at::Device device, bool non_blocking = false);
+
+  void synchronize(at::Device device);
+
+  void pin_memory();
+
+  void cuda_backup(bool non_blocking = true);
+
+  void clear();
+
   void save(
       std::ostream& out,
       const ExtraFilesMap& extra_files = ExtraFilesMap()) const;
diff --git a/torch/nn/modules/module.py b/torch/nn/modules/module.py
index 8fad7405450..f48f11bf104 100644
--- a/torch/nn/modules/module.py
+++ b/torch/nn/modules/module.py
@@ -421,7 +421,9 @@ class Module:
 
     training: bool
     _parameters: Dict[str, Optional[Parameter]]
+    _prev_parameters: Dict[str, Optional[Parameter]]
     _buffers: Dict[str, Optional[Tensor]]
+    _prev_buffers: Dict[str, Optional[Parameter]]
     _non_persistent_buffers_set: Set[str]
     _backward_pre_hooks: Dict[int, Callable]
     _backward_hooks: Dict[int, Callable]
@@ -472,7 +474,9 @@ class Module:
         """
         super().__setattr__('training', True)
         super().__setattr__('_parameters', OrderedDict())
+        super().__setattr__('_prev_parameters', OrderedDict())
         super().__setattr__('_buffers', OrderedDict())
+        super().__setattr__('_prev_buffers', OrderedDict())
         super().__setattr__('_non_persistent_buffers_set', set())
         super().__setattr__('_backward_pre_hooks', OrderedDict())
         super().__setattr__('_backward_hooks', OrderedDict())
@@ -1159,6 +1163,18 @@ class Module:
 
         return self._apply(convert)
 
+    def pin_memory(self, *args, **kwargs):
+        def convert(t):
+            return t if t.is_pinned() else t.pin_memory()
+
+        return self._apply(convert)
+
+    def cuda_host(self, *args, **kwargs):
+        def convert(t):
+            return t if t.is_pinned() else t.cuda_host()
+
+        return self._apply(convert)
+
     def register_full_backward_pre_hook(
         self,
         hook: Callable[["Module", _grad_t], Union[None, _grad_t]],
diff --git a/torchgen/native_function_generation.py b/torchgen/native_function_generation.py
index afd04b5fb01..95fb7b6a6a6 100644
--- a/torchgen/native_function_generation.py
+++ b/torchgen/native_function_generation.py
@@ -44,6 +44,8 @@ MUTABLE_OPS_THAT_CANNOT_GET_AN_OUT_VARIANT = [
     "_cummax_helper",
     # should be out=?
     "_cummin_helper",
+    "_record_event",
+    "_sync_event",
 ]
 
 # All of these operators don't have any tensor like returns
